### ISSUE TITLE
Get the top K recommendation for USERS

### ISSUE THREAD

**Issue by RayanAAY-ops** [pyg-team/pytorch_geometric#6614]
### 🚀 The feature, motivation and pitch

Hello community, I wanted to share a piece of code that helps me execute Neighbor Search among millions of candidates within a minute. I found this very easy and useful in a recommander system competition that needs to scale its solution.

In fact, I tried using the pyg documentation but couldn't easily retrieve the best candidates for each user. Here I want to propose a piece of code that allows me to do it in a Kaggle competition.
I applied it in the MovieLens dataset just like this :

```
!pip3 install annoy

%%time
from annoy import AnnoyIndex

knnMovies = AnnoyIndex(64, 'angular')

for idx,idx_embedding in enumerate(item_embedding):
    knnMovies.add_item(idx, idx_embedding)
    
knnMovies.build(10)


user_mapping = dict(zip(unique_user_id['userId'].values.tolist(),unique_user_id['mappedID'].values.tolist()))
movie_mapping = dict(zip(unique_movie_id['movieId'].values.tolist(),unique_movie_id['mappedID'].values.tolist()))

ratings_df['userId'] = ratings_df['userId'].map(user_mapping)
ratings_df['movieId'] = ratings_df['movieId'].map(movie_mapping)
users_movie_to_exclude = ratings_df.groupby("userId")['movieId'].apply(list)
k = 20 # We want to predict the next k best movie for each user and excluding the movie that has already been viewed by each of them.
predictions = []
for user_embed, index_ in zip(user_embedding,range(len(user_embedding))):
  nns = [i for i in knnMovies.get_nns_by_vector(user_embed, k)[:] if i not in users_movie_to_exclude[index_]]
  predictions.append(nns)

inference_df = pd.DataFrame(unique_user_id)
inference_df['labels'] = predictions
```

This piece of code give you the TOP K movies to recommend for each user directly delimited by a comma. I tried it with a graph of 7 millions users and 2 millions items, it runs pretty fast.

**Comment by rusty1s** [comment]
Thanks for the issue and the slack conversation :) What would be a good example to add this to?

**Comment by RayanAAY-ops** [comment]
> Thanks for the issue and the slack conversation :) What would be a good example to add this to?

I personally found it very useful for the MovieLens Recommandation task or any heterogenous graph architecture where we need to predict the next interactions.
In addition to being able to make the inference correctly by proposing the specific recommendations to each user ( space or comma delimited ), we can improve the validation step by adding a metric computation ( recall@K ... ). 

If you found it interesting, I can work on this last point quickly !

**Comment by rusty1s** [comment]
The current `examples/hetero/hetero_link_pred` example does utilize an MLP as decoder, so I think it is not directly applicable to this, right?

**Comment by RayanAAY-ops** [comment]
You're right. I propose a different approach based only on the embeddings extracted from the encoder
